pkg/storage/concurrency/concurrency_control.go

```diff
@@ -0,0 +1,355 @@
+	Timestamp hlc.Timestamp
+	// The priority of the request. Only set if Txn is nil.
+	Priority roachpb.UserPriority
+	// The consistency level of the request. Only set if Txn is nil.
+	ReadConsistency roachpb.ReadConsistencyType
+	// The individual requests in the batch.
+	Requests []roachpb.RequestUnion
+	// The maximal set of spans that the request will access.
+	Spans *spanset.SpanSet
+}
+
+// Guard is returned from Manager.SequenceReq. The guard is passed back in to
+// Manager.FinishReq to release the request's resources when it has completed.
+type Guard struct {
+	req  Request
+	lg   latchGuard
+	wqgs []lockWaitQueueGuard
+}
+
+// Response is a slice of responses to requests in a batch.
+type Response = []roachpb.ResponseUnion
+
+// Error is an alias for a roachpb.Error.
+type Error = roachpb.Error
+
+///////////////////////////////////
+// Internal Structure Interfaces //
+///////////////////////////////////
+
+// latchManager serializes access to keys and key ranges.
+//
+// See additional documentation in pkg/storage/spanlatch.
+type latchManager interface {
+	// Acquires latches, providing mutual exclusion for conflicting requests.
+	Acquire(context.Context, Request) (latchGuard, *Error)
+	// Releases latches.
+	Release(latchGuard)
+}
+
+// latchGuard is a handle to a set of acquired key latches.
+type latchGuard interface{}
+
+// lockTable holds a collection of locks acquired by in-progress transactions.
+// Each lock in the table has a possibly-empty lock wait-queue associated with
+// it, where conflicting transactions can queue while waiting for the lock to be
+// released.
+//
+//  +---------------------------------------------------+
+//  | [ lockTable ]                                     |
+//  | [    key1   ]    -------------+-----------------+ |
+//  | [    key2   ]  /  MVCCLock:   | lockWaitQueue:  | |
+//  | [    key3   ]-{   - lock type | <-[a]<-[b]<-[c] | |
+//  | [    key4   ]  \  - txn meta  |    (no latches) | |
+//  | [    key5   ]    -------------+-----------------+ |
+//  | [    ...    ]                                     |
+//  +---------------------------------------------------+
+//
+// TODO(nvanbenschoten): document further.
+type lockTable interface {
+	// AcquireLock informs the lockTable that a new lock was acquired or an
+	// existing lock was updated.
+	AcquireLock(roachpb.Intent)
+	// AcquireLock informs the lockTable that an existing lock was released.
+	ReleaseLock(roachpb.Intent)
+	// AddDiscoveredLock informs the lockTable of a lock that was discovered
+	// during evaluation that the lockTable wasn't previously tracking.
+	AddDiscoveredLock(Request, roachpb.Intent) lockWaitQueueGuard
+	// ScanAndEnqueue scans over the spans that the request will access and
+	// enqueues the request in the lock wait-queue of any conflicting locks
+	// encountered.
+	ScanAndEnqueue(Request, *Guard) []lockWaitQueueGuard
+	// Dequeue removes the guard from its lock wait-queue.
+	Dequeue(lockWaitQueueGuard)
```

Comment:
Is this called if a waiter gives up? In normal operation we'll only ever dequeue in FIFO order, right?

Reply:
Yes, this can be called if a waiter gives up, so we'll need to handle out of order dequeue-ing. Under normal operation though, you're correct. Commented.